File: src/main/java/org/gd/leetcode/p0240/Solution.py
# https://leetcode.com/problems/search-a-2d-matrix-ii/
class Solution:
    
     def searchMatrix(self, matrix, target: int):

        if not matrix or len(matrix) == 0 or len(matrix[0]) == 0:
            return False
        
        cols: int = len(matrix)
        rows: int = len(matrix[0])

        if (matrix[0][0] > target) or (matrix[-1][-1] < target):
            return False

        col: int = 0
        row: int = rows - 1

        while col < cols and row >= 0:

            if matrix[col][row] == target:
                return True

            if matrix[col][row] < target:
                col += 1
            else:
                row -= 1

        return False


matrix = [
  [1,   4,  7, 11, 15],
  [2,   5,  8, 12, 19],
  [3,   6,  9, 16, 22],
  [10, 13, 14, 17, 24],
  [18, 21, 23, 26, 30]
]

assert not Solution().searchMatrix([[]], 9)
assert Solution().searchMatrix(matrix, 9)
assert Solution().searchMatrix(matrix, 10)
assert not Solution().searchMatrix(matrix, 25)

File: src/main/java/org/gd/leetcode/p0240/Solution.java
package org.gd.leetcode.p0240;

import java.util.Arrays;

import org.gd.leetcode.common.Repeat;
import org.gd.leetcode.common.LeetCode;

/**
 * https://leetcode.com/problems/search-a-2d-matrix-ii/
 *
 * @see org.gd.leetcode.p0074.Solution
 * @see Arrays#binarySearch(int[], int)
 * @since 2019-09-10
 */
@SuppressWarnings("JavadocReference")
@Repeat("search value in 2d matrix")
@LeetCode(
        difficulty = LeetCode.Level.MEDIUM,
        tags = {
                LeetCode.Tags.BINARY_SEARCH,
                LeetCode.Tags.DIVIDE_AND_CONQUER
        })
class Solution {

    public boolean searchMatrix(int[][] matrix, int target) {

        if (matrix == null || matrix.length == 0)
            return false;

        int cols = matrix.length;
        int rows = matrix[0].length;

        if ((matrix[0][0] > target) || (matrix[cols - 1][rows - 1] < target))
            return false;

        for (int col = 0, row = rows - 1; col < cols && row >= 0;) {
            if (matrix[col][row] == target) {
                return true;
            }
            if (matrix[col][row] < target) {
                col++;
            } else {
                row--;
            }
        }

        return false;
    }
}
